lib/linkify.js

```diff
@@ -0,0 +1,84 @@
+// github-style linkification for linkify-it
+//
+// Pass in a markdown-it parser instance and this will return a version fixed
+// up for the kind of linkification we're after.
+//
+// As far as I can determine, the basic algorithm is to linkify anything
+// starting with 'www.' and ending with a) a whitespace character, or b) a dot
+// followed by a whitespace character (in other words, include internal dots,
+// but exclude a trailing dot). There are plenty of cases where this results in
+// invalid links, but that's how they do it, so that's how we'll do it.
+//
+
+function isInBrackets (text) {
+  return text.charAt(0) === '[' && text.charAt(text.length - 1) === ']'
+}
+
+function isInParens (text) {
+  return text.charAt(0) === '(' && text.charAt(text.length - 1) === ')'
+}
+
+module.exports = function (parser) {
+  if (parser) {
+    var linkify = parser.linkify
+
+    linkify.set({fuzzyLink: false}) // turn off auto-linking normal hostnames
+    linkify.add('//', null) // turn off protocol-relative links
+
+    // linkify everything hostnamey starting with 'www.', optionally in [] or ()
+    var scheme = {
+      validate: function (text, pos, self) {
+        if (!self.re.githubLinkify) {
+          self.re.githubLinkify = new RegExp('^' + self.re.src_host_port_strict + self.re.src_path, 'i')
+        }
+
+        // chop off any trailing ']' or ')', if necessary
+        var candidate = text.slice(pos)
+        if (isInBrackets(text) || isInParens(text)) {
+          candidate = candidate.slice(0, candidate.length - 1)
+        }
+
+        if (self.re.githubLinkify.test(candidate)) {
+          return candidate.match(self.re.githubLinkify)[0].length
+        } else if (candidate.charAt(0) === ' ') {
+          // if we have a zero-length match, i.e., if the original string was
+          // literally 'www.' all by itself, we return -1 because linkify-it does
+          // a boolean test on the result of the validator function (so we need a
+          // non-zero result here to make it through that test) and adds the
+          // returned value to an index it's using to point to the end of the
+          // matched substring. In this case, that means we'll move backwards one
+          // character and only link 'www' rather than the full 'www.' that was
+          // originally matched
+          return -1
+        }
+        return 0
+      },
+      normalize: function (match) {
+        // since we have 3 different "schemes" in place, our matched URLs might
+        // have a leading bracket or parenthesis; trim it off if necessary
+        if (match.schema.charAt(0) === 'w') {
+          match.url = 'http://' + match.url
+        } else {
+          match.url = 'http://' + match.url.slice(1)
+        }
+        match.schema = ''
+      }
+    }
+
+    parser.linkify.add('www.', scheme)
+    parser.linkify.add('[www.', scheme)
+    parser.linkify.add('(www.', scheme)
+
```

Comment:
I sort of feel like specifically adding "schemes" for the bracket- and parenthesis-enclosed versions of these is cheating, but it's the most convenient way of allowing just those characters through while also blocking other forms of punctuation.